codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java

```diff
@@ -0,0 +1,132 @@
+ * connection or didn't have a self defined message length.
+ * <p>
+ * Since {@link HttpServerKeepAliveHandler} expects {@link HttpObject}s it should be added after {@link HttpServerCodec}
+ * but before any other handlers that might send a {@link HttpResponse}. <blockquote>
+ * <pre>
+ *  {@link ChannelPipeline} p = ...;
+ *  ...
+ *  p.addLast("serverCodec", new {@link HttpServerCodec}());
+ *  p.addLast("httpKeepAlive", <b>new {@link HttpServerKeepAliveHandler}()</b>);
+ *  p.addLast("aggregator", new {@link HttpObjectAggregator}(1048576));
+ *  ...
+ *  p.addLast("handler", new HttpRequestHandler());
+ *  </pre>
+ * </blockquote>
+ */
+public class HttpServerKeepAliveHandler extends ChannelDuplexHandler {
+    private static final String MULTIPART_PREFIX = "multipart";
+
+    // Using null as a way to know if this is our first time setting keepAliveRequested to false
+    private Boolean keepAliveRequested;
+    // Track pending responses to support client pipelining: https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html
+    private int pendingResponses;
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        // read message and track if it was keepAlive
+        if (msg instanceof HttpRequest) {
+            final HttpRequest request = (HttpRequest) msg;
+            pendingResponses += 1;
+            if (!FALSE.equals(keepAliveRequested)) {
+                keepAliveRequested = isKeepAlive(request);
+            }
+        }
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        // modify message on way out to add headers if needed
+        if (msg instanceof HttpResponse) {
+            final HttpResponse response = (HttpResponse) msg;
+            trackResponse(response);
+            // Assume the response writer knows if they can persist or not and sets isKeepAlive on the response
+            if (!isKeepAlive(response) || !isSelfDefinedMessageLength(response)) {
+                // No longer keep alive as the client can't tell when the message is done unless we close connection
+                pendingResponses = 0;
+                keepAliveRequested = FALSE;
+            }
+            // Server might think it can keep connection alive, but we should fix response header if we know better
+            if (!shouldKeepAlive()) {
+                setKeepAlive(response, false);
+            }
+        }
+        if (msg instanceof LastHttpContent && !shouldKeepAlive()) {
+            promise.addListener(ChannelFutureListener.CLOSE);
+        }
+        super.write(ctx, msg, promise);
+    }
+
+    private void trackResponse(HttpResponse response) {
+        if (response.status().codeClass() != HttpStatusClass.INFORMATIONAL) {
+            pendingResponses -= 1;
+        }
+    }
+
+    private boolean shouldKeepAlive() {
+        if (pendingResponses == 0) {
```

Comment:
Would this be even more clear? Happy to make the change if you think it becomes even easier to read/understand. 

```java
return pendingResponses > 0 || persistentConnection;
```